ext/compressors/zstd/zstd_compress.c

```diff
@@ -0,0 +1,552 @@
+{
+	(void)session;					/* Unused parameters */
+
+	free(compressor);
+	return (0);
+}
+
+/*
+ * zstd_add_compressor --
+ *	Add a Zstd compressor.
+ */
+static int
+zstd_add_compressor(WT_CONNECTION *connection, int raw, const char *name)
+{
+	ZSTD_COMPRESSOR *zstd_compressor;
+
+	/*
+	 * There are two almost identical Zstd compressors: one using raw
+	 * compression to target a specific block size, and one without.
+	 */
+	if ((zstd_compressor = calloc(1, sizeof(ZSTD_COMPRESSOR))) == NULL)
+		return (errno);
+
+	zstd_compressor->compressor.compress = zstd_compress;
+	zstd_compressor->compressor.compress_raw = raw ?
+	    zstd_compress_raw : NULL;
+	zstd_compressor->compressor.decompress = zstd_decompress;
+	zstd_compressor->compressor.pre_size = zstd_pre_size;
+	zstd_compressor->compressor.terminate = zstd_terminate;
+
+	zstd_compressor->wt_api = connection->get_extension_api(connection);
+
+	/*
+	 * Zstd's sweet-spot is better compression than zlib at significantly
+	 * faster compression/decompression speeds. LZ4 and snappy are faster
+	 * than zstd, but have worse compression ratios. Applications wanting
+	 * faster compression/decompression with worse compression will select
+	 * LZ4 or snappy, so we configure zstd for better compression.
+	 *
+	 * From the zstd github site, default measurements of the compression
+	 * engines we support, listing compression ratios with compression and
+	 * decompression speeds:
+	 *
+	 *	Name	Ratio	C.speed	D.speed
+	 *			MB/s	MB/s
+	 *	zstd	2.877	330	940
+	 *	zlib	2.730	95	360
+	 *	LZ4	2.101	620	3100
+	 *	snappy	2.091	480	1600
+	 *
+	 * Set the zstd compression level to 3: according to the zstd web site,
+	 * that reduces zstd's compression speed to around 200 MB/s, increasing
+	 * the compression ratio to 3.100 (close to zlib's best compression
+	 * ratio). In other words, position zstd as a zlib replacement, having
+	 * similar compression at much higher compression/decompression speeds.
+	 */
+	zstd_compressor->compression_level = 3;
+
+	/*
+	 * Experimentally derived, reserve this many bytes for zlib to finish
+	 * up a buffer. If this isn't sufficient, we don't fail but we will be
+	 * inefficient.
+	 */
+	zstd_compressor->finish_reserve = 256;
+
+	/* Load the compressor */
+	return (connection->add_compressor(
+	    connection, name, (WT_COMPRESSOR *)zstd_compressor, NULL));
+}
+
+int zstd_extension_init(WT_CONNECTION *, WT_CONFIG_ARG *);
+
+/*
+ * zstd_extension_init --
+ *	WiredTiger Zstd compression extension - called directly when Zstd
+ * support is built in, or via wiredtiger_extension_init when Zstd support
+ * is included via extension loading.
+ */
+int
+zstd_extension_init(WT_CONNECTION *connection, WT_CONFIG_ARG *config)
+{
+	int ret;
+
+	(void)config;    				/* Unused parameters */
+
+	if ((ret = zstd_add_compressor(connection, 1, "zstd")) != 0)
+		return (ret);
+	if ((ret = zstd_add_compressor(connection, 0, "zstd-noraw")) != 0)
+		return (ret);
```

Comment:
As discussed, let's at least flip this so we have the default "zstd" not include raw compression (matching snappy).  If testing suggests that the overhead of "zstd-raw" is justified given the on-disk space saving, we can leave it in, otherwise let's remove that code.